pkg/cloud/services/compute/instance.go

```diff
@@ -303,24 +335,95 @@ func getPortName(instanceName string, opts *infrav1.PortOpts, netIndex int) stri
 	return fmt.Sprintf("%s-%d", instanceName, netIndex)
 }
 
-// applyRootVolume sets a root volume if the root volume Size is not 0.
-func applyRootVolume(opts servers.CreateOptsBuilder, rootVolume *infrav1.RootVolume) servers.CreateOptsBuilder {
-	if rootVolume != nil && rootVolume.Size != 0 {
-		block := bootfromvolume.BlockDevice{
-			SourceType:          bootfromvolume.SourceType(rootVolume.SourceType),
-			BootIndex:           0,
-			UUID:                rootVolume.SourceUUID,
-			DeleteOnTermination: true,
-			DestinationType:     bootfromvolume.DestinationVolume,
-			VolumeSize:          rootVolume.Size,
-			DeviceType:          rootVolume.DeviceType,
-		}
-		return bootfromvolume.CreateOptsExt{
-			CreateOptsBuilder: opts,
-			BlockDevice:       []bootfromvolume.BlockDevice{block},
+func rootVolumeName(instanceName string) string {
+	return fmt.Sprintf("%s-root", instanceName)
+}
+
+func hasRootVolume(rootVolume *infrav1.RootVolume) bool {
+	return rootVolume != nil && rootVolume.Size > 0
+}
+
+func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
+	listOpts := volumes.ListOpts{
+		AllTenants: false,
+		Name:       name,
+		TenantID:   s.projectID,
+	}
+	volumeList, err := s.computeService.ListVolumes(listOpts)
+	if err != nil {
+		return nil, fmt.Errorf("error listing volumes: %w", err)
+	}
+	if len(volumeList) > 1 {
+		return nil, fmt.Errorf("expected to find a single volume called %s; found %d", name, len(volumeList))
+	}
+	if len(volumeList) == 0 {
+		return nil, nil
+	}
+	return &volumeList[0], nil
+}
+
+func (s *Service) getOrCreateRootVolume(eventObject runtime.Object, instanceSpec *InstanceSpec, imageID string) (*volumes.Volume, error) {
+	rootVolume := instanceSpec.RootVolume
+	if !hasRootVolume(rootVolume) {
+		return nil, nil
+	}
+
+	name := rootVolumeName(instanceSpec.Name)
+	size := rootVolume.Size
+
+	volume, err := s.getVolumeByName(name)
+	if err != nil {
+		return nil, err
+	}
+	if volume != nil {
+		if volume.Size != size {
+			return nil, fmt.Errorf("exected to find volume %s with size %d; found size %d", name, size, volume.Size)
 		}
+
+		s.logger.Info("using existing root volume %s", name)
+		return volume, nil
```

Comment:
This returns a volume selected by name and size only. I wonder if there are two volumes in different availability zones or volume types, I am not sure that could happen though.

Reply:
Firstly, I agree. It's hopefully unlikely in practise, but I'd really like to eliminate everywhere that we rely on resource name because it introduces the potential for a naming conflict, and it's less efficient for the API.

[As mentioned elsewhere](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1095#pullrequestreview-848928494) I'd also like to clean up how this works to simplify the resource creation/deletion logic.

My current idea for how to do this is to track explicitly-created cloud resources in Status. This would look something like:

```
    function ReconcileRootVolume(osMachine) (reconciled bool) {
        if osMachine.Status.RootVolume != "" {
            return volumeIsAvailable(rootVolumeUUID)
        }
        
        osMachine.Status.RootVolume = createRootVolume()
        return false
    }
```

This also makes cleanup more robust because you know you have to delete a root volume [iff](https://en.wikipedia.org/wiki/If_and_only_if) RootVolume is set.

The same would also be true of ports. It would also mean we could create ports while waiting for the root volume to become available by moving polling entirely out of the create function, and simply asking the controller to call reconcile again in 10 seconds.

I was *extremely* tempted to do this in this patch, but I didn't because this patch is already big enough, and already makes a significant API change in the Spec. I very much want to come back to it, though.

In the meantime I don't think it's possible to avoid doing this match on name because that implies we'd have to store a uuid somewhere, and we don't yet have anywhere to put it.

Reply:
Thanks for not clean up creation/deletion logic at the same time for reviewers 😃 I agree with your idea using uuid in Status to identify resources.